File: scala/parsers/src/main/scala/jp/co/teruuu/parser/common/ManyParser.scala
package jp.co.teruuu.parser.common

class ManyParser[T](parser: Parser[T]) extends Parser[List[T]] {

  override def parse(input: String): ParseResult[List[T]] = {
    var next = input
    var values = List.empty[T]
    var isContinue = true
    while(isContinue) {
      val result = parser.parse(next)
      if(result.isInstanceOf[ParseOk[_]]) {
        values = values :+ result.asInstanceOf[ParseOk[T]].value
        next = result.asInstanceOf[ParseOk[T]].next
      } else {
        isContinue = false
      }
    }
    ParseOk(values, next)
  }
}

File: java/parsers/src/main/java/jp/co/teruuu/parser/common/ManyParser.java
package jp.co.teruuu.parser.common;

import java.util.ArrayList;
import java.util.List;

/**
 * 0個以上の連続をパースする
 *
 * @param <T>
 */
public class ManyParser<T> implements Parser<List<T>> {
    private Parser<T> parser;

    public ManyParser(Parser<T> parser) {
        this.parser = parser;
    }

    @Override
    public ParseResult<List<T>> parse(String input) {
        ParseResult<T> result;
        String next = input;
        List<T> values = new ArrayList<T>();
        while (true) {
            String previous = next;
            result = parser.parse(next);
            if (result instanceof ParseResult.Success<?>) {
                ParseResult.Success<T> success = (ParseResult.Success<T>) result;
                values.add(success.value);
                next = success.next;
            } else {
                next = null;
            }
            if (next == null) {
                return new ParseResult.Success<>(values, previous);
            }
        }
    }
}